src/components/Personalization/helper/dom/matchesSelector.js

```diff
@@ -0,0 +1,31 @@
+/*
+Copyright 2019 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import { selectNodesWithEq } from "./selectNodesWithEq";
+
+// Using node selection vs matches selector, because of :eq()
+export default (selector, element) => {
+  // Find all nodes using document as context
+  const nodes = selectNodesWithEq(selector, document);
```

Comment:
@jonsnyder yeah, I saw that this might not be optimal, but I wanted to make sure that it is OK functional wise.

If this will become a HUGE issue we can review and come up with a better algo. The first that comes to mind is that instead of invoking all that selector parsing we could memoize or do something smarter.

Thanks for you suggestion, I'll add it to my next commit.

Reply:
Hey @artur-ciocanu, I don't think you've really resolved what @jonsnyder and I were pointing out. You may be intentionally putting it off until later, but I just want to point in out in case it was an oversight.

Considering 20 elements in the hierarchy of the clicked element and 10 items in storage (all using `:eq`), your later commit changed the logic from:

```
for each element in clicked element hierarchy: 20 elements
  for each item in storage: 10 items
    for each found element that matches selectNodesWithEq: 200 selectNodesWithEq queries
      check to to see if found element == current element in hierarchy
```

to

```
for each item in storage: 10 items
  for each element in clicked element hierarchy: 20 elements
    for each found element that matches selectNodesWithEq: 200 selectNodesWithEq queries
      check to to see if found element == current element in hierarchy
```

In both cases, there are 200 calls to `selectNodesWithEq`, which I think is the concern because it's likely to be the bottleneck.

What was being suggested, I think, was that you re-work the logic to only call `selectNodesWithEq` 10 times (one for each item in storage), like this:

```
for each item in storage, call selectNodesWithEq: 10 selectNodesWithEq queries
  for each element in clicked element hierarchy: 20 elements
    check to see if element in hierarchy is in selectNodesWithEq result: this occurs 200 times total
```

Note that your current approach will be fastest when `Element.matches` can be used (the selector doesn't have `:eq`), whereas the proposed approach will be fastest otherwise. You might have to have a split code path here.

If you end up not implementing the proposed approach, I think you can optimize your current approach by changing this:

```
const nodes = selectNodesWithEq(selector, document);
```

to this:

```
const nodes = selectNodesWithEq(selector, element.parentNode);
```

So that you're querying within a smaller scope than `document`. However, that requires `selectNodesWithEq()` to work properly when you pass in a scope other than `document`, and I don't think it is right now. You can see this if you run this test inside `selectNodesWithEq.spec.js`:

```
it("should select when eq and just one element", () => {
    const content = `
      <div class="b">
        <div class="c">first</div>

        <div class="c">second</div>
        
        <div class="c">third</div>
      </div>
    `;

    appendNode(
      document.body,
      createNode("DIV", { id: "abc", class: "eq" }, { innerHTML: content })
    );

    const result = selectNodesWithEq(
      "#abc:eq(0) > div.b:eq(0) > div.c:eq(0)",
      document.getElementById("abc")
    );

    expect(result[0].tagName).toEqual("DIV");
    expect(result[0].textContent).toEqual("first");
  });
```

This fails to find an element. However, if you switch out:

```
const result = selectNodesWithEq(
      "#abc:eq(0) > div.b:eq(0) > div.c:eq(0)",
      document.getElementById("abc")
    );
```

with

```
const result = document
      .getElementById("abc")
      .querySelectorAll("#abc > div.b > div.c");
```

you'll see that it does find a result. This should either get fixed (might be super tricky) or remove the context parameter in `selectNodesWithEq`.